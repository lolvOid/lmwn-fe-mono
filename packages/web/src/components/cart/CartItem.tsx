import { getThaiBaht } from '@/utilts/formatCurrency';
import { GrTrash } from 'react-icons/gr';
import { CartDataType } from '../../types/cartTypes';

interface CartItemProps {
    itemData: CartDataType;
    onRemove?: (id: number | string) => void;
}
const CartItem = ({ itemData, onRemove }: CartItemProps) => {
    return (
        <>
            <div className="flex justify-between p-2 flex-col">
                <div className="flex">
                    <img
                        src={
                            itemData.thumbnailImage ||
                            'https://dummyimage.com/128x128/dedede/5e5e5e.jpg&text=No+Image+Available'
                        }
                        className="w-16 h-16 object-cover object-center flex-shrink-0"
                        alt={itemData.name || 'Item'}
                    />
                    <span className="primary-text text-xl ml-3">{itemData.name}</span>
                    <span className="text-md ml-3">
                        {getThaiBaht(itemData.fullPrice * itemData.count)}
                    </span>
                </div>
                {itemData.options.length > 0 && (
                    <>
                        <div className="flex justify-between">
                            <span className="mx-auto  text-center w-16 text-xs">
                                {itemData.count}
                            </span>
                            <div className="flex flex-col flex-grow px-3 text-wrap">
                                <span className="text-xs text-gray-500 flex flex-col">
                                    {itemData.options.map((option, index) => {
                                        return <span key={index}>{option}</span>;
                                    })}
                                </span>
                            </div>
                        </div>
                    </>
                )}
                <button
                    className="border-2 border-gray-400  text-gray-400 p-2 mt-2 rounded-md hover:bg-gray-400 hover:text-white transition-all inline-flex items-center justify-center"
                    onClick={() => {
                        onRemove && onRemove(itemData.id);
                    }}
                >
                    <GrTrash className="mr-2" />
                    Remove
                </button>
            </div>
        </>
    );
};

export default CartItem;
