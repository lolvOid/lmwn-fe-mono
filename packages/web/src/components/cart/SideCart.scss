.side-cart {
    height: calc(100vh - 4rem);
}