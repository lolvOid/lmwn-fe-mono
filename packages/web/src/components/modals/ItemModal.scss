
.item-modal {
    .item-modal-content {
        scrollbar-width: none;
    }
    .choices-checkbox {
        label {
            display: block;
            position: relative;
            font-size: 14px !important;
            cursor: pointer;
            color: black;
            -webkit-user-select: none;
            -moz-user-select: none;
            -ms-user-select: none;
            user-select: none;
        }
        [type='radio'] {
            border-radius: 2px;
        }
        [type='checkbox']:checked, [type='radio']:checked{
            background-size: 10px;
            background-color: #ffb700;
            background-image: url("data:image/svg+xml;charset=UTF-8,%3csvg stroke='currentColor' fill='%23ffffff' stroke-width='0' viewBox='0 0 512 512' height='1em' width='1em' xmlns='http://www.w3.org/2000/svg'%3e%3cpath d='M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z'%3e%3c/path%3e%3c/svg%3e");
        }
        [type='checkbox']:focus, [type='radio']:focus {
            outline: none !important;
            box-shadow: none;
            outline-offset: 0 !important;
        }
    
    }
    .cart-button:disabled {
        background-color: gray;
    }
}