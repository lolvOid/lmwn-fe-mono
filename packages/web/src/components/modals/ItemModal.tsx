import { useEffect, useState, Fragment, BaseSyntheticEvent } from 'react';
import { FaMinus, FaPlus } from 'react-icons/fa';
import { GrClose } from 'react-icons/gr';
import useModalStore from '@/store/modal/modalStore';
import useFullMenuItemQuery from '@/services/queries/menu.query';
import { FullMenu } from '@/interfaces/menu';
import { getThaiBaht } from '@/utils/formatCurrency';
import { getDiscountedPrice } from '@/utils/helper';
import useCartStore from '@/store/cart/cartStore';
import styles from '@/components/modals/ItemModal.module.scss';
import { CartDataType } from '@/store/types/cartTypes';
interface ItemModalProps {
    restaurantId: number | string;
}

const ItemModal = ({ restaurantId }: ItemModalProps) => {
    const { show, modalData, hideModal, resetModalData } = useModalStore();
    const [cartCount, setCartCount] = useState(1);
    const [selectedOptions, setOptions] = useState<string[]>([]);
    const { cartData, addToCart, calculateTotalCartPrice, calculateTotalCartCount } =
        useCartStore();
    const { data, isLoading, isError } = useFullMenuItemQuery(
        modalData?.restaurantId || '',
        modalData?.menuName || '',
        'full',
    );
    const [itemData, setItemData] = useState<FullMenu>();
    const handleAddToCart = (value: CartDataType) => {
        addToCart(value);
        hideModal();
        calculateTotalCartPrice();
        calculateTotalCartCount();
    };
    const handleIncreaseCartCount = () => {
        setCartCount(cartCount + 1);
    };

    const handleReduceCartCount = () => {
        if (cartCount <= 1) {
            setCartCount(1);
        } else {
            setCartCount(cartCount - 1);
        }
    };
    const handleOptions = (event: BaseSyntheticEvent) => {
        const selectedValue = event.target?.value || '';
        const isValueSelected = selectedOptions.includes(selectedValue);
        if (isValueSelected) {
            const updatedOptions: string[] = selectedOptions.filter(
                (option) => option !== selectedValue,
            );
            setOptions(updatedOptions);
        } else {
            const updatedOptions: string[] = [...selectedOptions, selectedValue];
            setOptions(updatedOptions);
        }
    };
    useEffect(() => {
        if (show) {
            if (isError) {
                console.error('Error fetching menu data:', isError);
            }

            if (!isLoading && data) {
                setItemData(data);
            }
        }
    }, [show, isLoading, isError, data, cartData]);
    return (
        <>
            {show && itemData && Object.keys(itemData).length && (
                <div
                    className={`fixed top-0 left-0 w-screen h-screen z-[1000]  bg-opacity-60 bg-black ${styles.itemModal}`}
                >
                    <div className="w-full h-full flex justify-center items-end lg:items-center">
                        <div
                            className={`relative rounded-t-2xl lg:rounded-lg shadow-md shadow-gray-400 flex flex-col z-[2000] lg:w-[600px] h-[80vh] lg:max-h-[1024px] w-full overflow-y-auto max-w-screen-2xl max-h-svh bg-white ${styles.itemModalContent}`}
                        >
                            <div className="w-full rounded-t-2xl lg:rounded-t-lg">
                                {!itemData?.largeImage && (
                                    <div
                                        role="status"
                                        className="space-y-8 animate-pulse md:space-y-0 md:space-x-8 rtl:space-x-reverse md:flex md:items-center"
                                    >
                                        <div className="flex items-center justify-center w-full h-48 bg-gray-300 rounded-t-2xl lg:rounded-t-lg">
                                            <svg
                                                className="w-full h-[250px] text-gray-200"
                                                aria-hidden="true"
                                                xmlns="http://www.w3.org/2000/svg"
                                                fill="currentColor"
                                                viewBox="0 0 20 18"
                                            >
                                                <path d="M18 0H2a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2Zm-5.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm4.376 10.481A1 1 0 0 1 16 15H4a1 1 0 0 1-.895-1.447l3.5-7A1 1 0 0 1 7.468 6a.965.965 0 0 1 .9.5l2.775 4.757 1.546-1.887a1 1 0 0 1 1.618.1l2.541 4a1 1 0 0 1 .028 1.011Z" />
                                            </svg>
                                        </div>
                                    </div>
                                )}
                                {itemData?.largeImage && (
                                    <img
                                        src={itemData?.largeImage}
                                        className="object-cover w-full h-[250px] rounded-t-2xl lg:rounded-t-lg"
                                        alt={itemData.name}
                                    />
                                )}
                            </div>
                            <button
                                data-testid="close-button"
                                className="w-auto h-auto p-4 absolute top-1 right-1 hover:bg-white hover:bg-opacity-50 rounded-full"
                                onClick={() => {
                                    hideModal();
                                    resetModalData();
                                }}
                            >
                                <GrClose size={24} />
                            </button>
                            <div className="flex flex-col gap-2 flex-grow p-4 relative">
                                <div className="border-b-2 h-fit py-4 items-end inline-flex z-10 bg-white w-full justify-between sticky top-0">
                                    <span
                                        className="text-xl font-bold pr-4"
                                        data-testid="item-name"
                                    >
                                        {itemData.name}
                                    </span>
                                    <span className="inline text-2xl font-light">
                                        <span>
                                            {itemData.discountedPercent
                                                ? getThaiBaht(
                                                      getDiscountedPrice(
                                                          itemData.fullPrice,
                                                          itemData.discountedPercent,
                                                      ),
                                                  )
                                                : getThaiBaht(itemData.fullPrice)}
                                        </span>
                                        <span className="pl-2 text-xl text-gray-500 line-through">
                                            {itemData.discountedPercent !== 0 &&
                                                getThaiBaht(itemData.fullPrice)}
                                        </span>
                                    </span>
                                </div>
                                <div className="mt-4 h-fit py-2 flex flex-col w-full justify-between">
                                    {itemData.options?.length > 0 &&
                                        itemData.options.map((option, index) => {
                                            return (
                                                <Fragment key={index}>
                                                    <span className="text-md font-bold my-3">
                                                        {option.label}
                                                    </span>
                                                    <span className="flex flex-col gap-2 text-sm">
                                                        {option.choices?.length &&
                                                            option.choices.map((choice, index) => {
                                                                return (
                                                                    <div
                                                                        key={index}
                                                                        className={`inline-flex gap-2 items-center ${styles.choicesCheckbox}`}
                                                                    >
                                                                        <input
                                                                            type="checkbox"
                                                                            name={option.label}
                                                                            id={choice.label}
                                                                            value={choice.label}
                                                                            onChange={handleOptions}
                                                                        />
                                                                        <label
                                                                            htmlFor={choice.label}
                                                                            className="text-md"
                                                                        >
                                                                            {choice.label}
                                                                        </label>
                                                                    </div>
                                                                );
                                                            })}
                                                    </span>
                                                </Fragment>
                                            );
                                        })}
                                </div>
                            </div>
                            <div className="flex p-4 border-t-2 gap-8 items-center justify-between sticky bottom-0 left-0 w-full bg-white">
                                <div className="flex">
                                    <button
                                        className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-l-md hover:bg-yellow-500"
                                        onClick={() => handleReduceCartCount()}
                                    >
                                        <FaMinus />
                                    </button>
                                    <span className="w-8 h-8 inline-flex text-md items-center justify-center border-y-2 border-yellow-300 border-solid">
                                        {cartCount}
                                    </span>
                                    <button
                                        className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-r-md rounded-l-none rounded-md hover:bg-yellow-500"
                                        onClick={() => handleIncreaseCartCount()}
                                    >
                                        <FaPlus />
                                    </button>
                                </div>
                                <div className="flex-grow">
                                    <button
                                        disabled={
                                            itemData.options.length > 0 &&
                                            selectedOptions.length === 0
                                        }
                                        data-testid="add-to-cart-button"
                                        className="primary-bg text-md w-full h-10 inline-flex items-center disabled:bg-gray-300 justify-center text-white rounded-md hover:bg-yellow-500"
                                        onClick={() =>
                                            handleAddToCart({
                                                id: itemData.id,
                                                name: itemData.name,
                                                count: cartCount,
                                                restaurantId: restaurantId,
                                                fullPrice: itemData.fullPrice,
                                                options: selectedOptions,
                                                thumbnailImage: itemData.thumbnailImage,
                                            })
                                        }
                                    >
                                        <span>Add to Cart</span>
                                    </button>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            )}
        </>
    );
};

export default ItemModal;
