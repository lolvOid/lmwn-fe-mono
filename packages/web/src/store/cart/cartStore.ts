import { create } from 'zustand';

export type CartDataType = {
    id: number | string;
    name: string | undefined;
    restaurantId: number|string;
    fullPrice: number;
    count: number;
    options: string[],
    thumbnailImage: string | undefined,
};

type CartStore = {
    isCartDisplayed: boolean;
    cartData: CartDataType[];
    totalCartPrice: number;
    totalCartCount: number;
    showCart: () => void;
    hideCart: () => void;
    addToCart: (value: CartDataType) => void;
    resetCart: () => void;
    removeFromCartById:(id: number | string) => void;
    increaseCount: (id: number | string) => void;
    decreaseCount: (id: number | string) => void;
    updateOptions: (id: number | string, options: string[]) => void;
    updateCartCountById: (id: number | string, count: number) => void;
    calculateTotalCartPrice: () => void;
    calculateTotalCartCount: () => void;
};

const useCartStore = create<CartStore>((set) => ({
    isCartDisplayed: false,
    cartData: [],
    totalCartPrice: 0,
    totalCartCount: 0,
    addToCart: (value: CartDataType) => set((state) => {
        const existingItemIndex = state.cartData.findIndex((item) => item.id === value.id);
        if (existingItemIndex !== -1) {
            const updatedCartData = state.cartData.map((item, index) => {
                if (index === existingItemIndex) {
                    return {
                        ...item,
                        name: item.name,
                        count: item.count + value.count,
                        options: [...item.options, ...value.options],
                        fullPrice: item.fullPrice,
                        thumbnailImage: item.thumbnailImage,
                    };
                }
                return item;
            });
            return { ...state, cartData: updatedCartData };
        } else {
            return { ...state, cartData: [...state.cartData, value] };
        }
    }),
    increaseCount: (id: number | string) => set((state) => {
        const cartItemIndex = state.cartData.findIndex((item) => item.id === id);
        if (cartItemIndex !== -1) {
            const updatedCartData = state.cartData.map((item, index) => {
                if (index === cartItemIndex) {
                    return { ...item, count: item.count + 1 };
                }
                return item;
            });
            return { ...state, cartData: updatedCartData };
        }
        return state;
    }),
    decreaseCount: (id: number | string) => set((state) => {
        const cartItemIndex = state.cartData.findIndex((item) => item.id === id);
        if (cartItemIndex !== -1) {
            const updatedCartData = state.cartData.map((item, index) => {
                if (index === cartItemIndex) {
                    return { ...item, count: item.count + 1 };
                }
                return item;
            });
            return { ...state, cartData: updatedCartData };
        }
        return state;
    }),
    updateOptions: (id: number | string, options: string[]) => set((state) => {
        const cartItemIndex = state.cartData.findIndex((item) => item.id === id);
        if (cartItemIndex !== -1) {
            const updatedCartData = state.cartData.map((item, index) => {
                if (index === cartItemIndex) {
                    return { ...item, options };
                }
                return item;
            });
            return { ...state, cartData: updatedCartData };
        }
        return state;
    }),
    removeFromCartById: (id: number | string) => set((state) => ({
        ...state,
        cartData: state.cartData.filter((item) => item.id !== id),
    })),
    showCart: () => set({ isCartDisplayed: true }),
    hideCart: () => set({ isCartDisplayed: false }),
    updateCartCountById: (id: number | string, count: number) => set((state) => {
        const cartItemIndex = state.cartData.findIndex((item) => item.id === id);
        if (cartItemIndex !== -1) {
            const updatedCartData = state.cartData.map((item, index) => {
                if (index === cartItemIndex) {
                    return { ...item, count: count };
                }
                return item;
            });
            return { ...state, cartData: updatedCartData };
        }
        return state;
    }),
    resetCart: () => set({ cartData: [] }),
    calculateTotalCartPrice: () => set((state) => {
        const VAT = 7;
        const totalCartPriceWithoutVAT = state.cartData.reduce((total, item) => total + (item.fullPrice * item.count), 0);
        const VATAmount = (VAT / 100) * totalCartPriceWithoutVAT;
        const totalCartPrice = totalCartPriceWithoutVAT + VATAmount;
        return { ...state, totalCartPrice };
    }),
    calculateTotalCartCount: () => set((state) => {
        const totalCartCount = state.cartData.reduce((totalCount, item) => totalCount + item.count, 0);
        return { ...state, totalCartCount };
    }),

}));

export default useCartStore;