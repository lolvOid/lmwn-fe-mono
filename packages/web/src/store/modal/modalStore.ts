import { ModalDataType, ModalStore } from '@/types/modalTypes';
import { create } from 'zustand';

const useModalStore = create<ModalStore>((set) => ({
    show: false,
    modalData: {
        menuName: '',
        restaurantId: '',
    },
    setModalData: (value: ModalDataType) => set({ modalData: value }),
    showModal: () => set({ show: true }),
    hideModal: () => set({ show: false }),
    resetModalData: () =>
        set({
            modalData: {
                menuName: '',
                restaurantId: '',
            },
        }),
}));

export default useModalStore;
