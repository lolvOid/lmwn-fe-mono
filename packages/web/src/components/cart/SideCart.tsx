import useCartStore from '@/store/cart/cartStore';
import CartItem from '@/components/cart/CartItem';
import { getThaiBaht } from '@/utilts/formatCurrency';
const SideCart = () => {
    const {
        cartData,
        totalCartPrice,
        isCartDisplayed,
        resetCart,
        hideCart,
        removeFromCartById,
        calculateTotalCartPrice,
    } = useCartStore();
    const handleOrder = () => {
        alert('Order Confirmed');
        resetCart();
        calculateTotalCartPrice();
        hideCart();
    };
    const handleRemoveItem = (id: number | string) => {
        removeFromCartById(id);
        calculateTotalCartPrice();
    };
    return (
        <>
            {isCartDisplayed && (
                <div
                    className="fixed w-screen md:lg:w-[400px] bg-white top-16 right-0 z-[500] shadow-md shadow-neutral-300  overflow-y-auto"
                    style={{
                        height: 'calc(100vh - 4rem)',
                    }}
                >
                    <div className="border-2 rounded-md flex m-2 p-3 flex-col h-fit">
                        {cartData.length === 0 && (
                            <span className="text-center">No items to display</span>
                        )}
                        {cartData.length > 0 &&
                            cartData.map((cartItem, index) => {
                                return (
                                    <CartItem
                                        itemData={cartItem}
                                        key={index}
                                        onRemove={handleRemoveItem}
                                    />
                                );
                            })}
                    </div>
                    <div
                        className={`bg-white p-4 bottom-0 right-0 flex flex-col justify-center items-center `}
                    >
                        <div className="inline-flex w-full justify-between py-4">
                            <span className="text-left ">
                                Total <span className="text-xs text-gray-600">(incl VAT):</span>
                            </span>
                            <span className="text-right">{getThaiBaht(totalCartPrice)}</span>
                        </div>
                        <button
                            className="p-4 flex text-center items-center justify-center disabled:bg-gray-300 primary-bg text-white  rounded-lg w-full"
                            onClick={handleOrder}
                            disabled={cartData.length === 0}
                        >
                            Confirm and Place Order
                        </button>
                    </div>
                </div>
            )}
        </>
    );
};
export default SideCart;
