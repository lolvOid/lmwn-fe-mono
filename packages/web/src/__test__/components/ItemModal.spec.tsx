import { render, screen, fireEvent } from '@testing-library/react';
import ItemModal from '@/components/modals/ItemModal';
import * as modalStoreModule from '@/store/modal/modalStore';
import * as cartStoreModule from '@/store/cart/cartStore';
import useFullMenuItemQuery from '@/services/queries/menu.query';

jest.mock('@/services/queries/menu.query');

describe('ItemModal', () => {
    const mockHideModal = jest.fn();
    const mockResetModalData = jest.fn();
    const mockAddToCart = jest.fn();

    beforeEach(() => {
        jest.spyOn(modalStoreModule, 'default').mockReturnValue({
            show: true,
            modalData: {
                restaurantId: 227018,
                menuName: 'Egg',
            },
            setModalData: jest.fn(),
            showModal: jest.fn(),
            hideModal: mockHideModal,
            resetModalData: mockResetModalData,
        });

        jest.spyOn(cartStoreModule, 'default').mockReturnValue({
            addToCart: mockAddToCart,
            calculateTotalCartPrice: jest.fn(),
            calculateTotalCartCount: jest.fn(),
        });

        (useFullMenuItemQuery as jest.Mock).mockReturnValue({
            data: {
                name: 'Egg',
                id: 'Egg',
                thumbnailImage: 'thumb.jpg',
                fullPrice: 200,
                discountedPercent: 0,
                discountedTimePeriod: {
                    begin: '0',
                    end: '0',
                },
                sold: 200,
                totalInStock: 10,
                largeImage: 'sample.jpg',
                options: [],
            },
            isLoading: false,
            isError: false,
        });
    });

    it('Should render properly with provided modalData', async () => {
        render(<ItemModal restaurantId={227018} />);
        await screen.findByTestId('item-name');
        expect(screen.getByTestId('item-name')).toHaveTextContent('Egg');
        fireEvent.click(screen.getByTestId('close-button'));
        expect(mockHideModal).toHaveBeenCalled();
        expect(mockResetModalData).toHaveBeenCalled();
    });

    it('Should handle adding to cart properly', async () => {
        render(<ItemModal restaurantId={227018} />);
        const addToCartButton = screen.getByTestId('add-to-cart-button');
        fireEvent.click(addToCartButton);
        expect(mockAddToCart).toHaveBeenCalledWith({
            id: 'Egg',
            name: 'Egg',
            count: 1,
            restaurantId: 227018,
            fullPrice: 200,
            options: [],
            thumbnailImage: 'thumb.jpg',
        });
    });
});
