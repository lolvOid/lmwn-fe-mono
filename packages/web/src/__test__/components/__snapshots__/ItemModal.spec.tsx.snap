// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemModal Should render and match the default snapshot 1`] = `null`;
